calico_containers/pycalico/datastore_datatypes.py

```diff
@@ -396,6 +396,14 @@ def __setitem__(self, key, value):
             value = IPNetwork(value)
         if key == "action" and value not in ("allow", "deny"):
             raise ValueError("'%s' is not allowed for key 'action'" % value)
+        if key == "protocol" and value not in ("tcp", "udp", "icmp", None):
+            raise ValueError("'%s' is not allowed for key 'protocol'" % value)
+        if key in ("src_tag", "dst_tag") and re.search('[^a-zA-Z0-9\.\_\-]', value):
```

Comment:
You don't need to escape any of the last characters do you?
. should mean "." when in [ ]
I don't think _ means anything
and the - should be fine when at the beginning or the end of the list of chars

might be wrong tho.